client/src/main/java/io/pravega/client/control/impl/ControllerImpl.java

```diff
@@ -1282,30 +1302,78 @@ private StreamSegments getStreamSegments(final SegmentRanges ranges, long reques
     public CompletableFuture<PravegaNodeUri> getEndpointForSegment(final String qualifiedSegmentName) {
         Exceptions.checkNotClosed(closed.get(), this);
         Exceptions.checkNotNullOrEmpty(qualifiedSegmentName, "qualifiedSegmentName");
+        Segment segment = Segment.fromScopedName(qualifiedSegmentName);
+        CachedPravegaNodeUri nodeUri = getSegmentEndpointFromCache(segment);
+        //Read from cache if the segment endpoint already exists and the refresh interval is not expired.
+        if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() <= CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            log.info("Fetching the endpoint details of segment {} from cache", segment.getSegmentId());
+            return nodeUri.getPravegaNodeUri();
+        } else if (nodeUri != null && nodeUri.getTimer().getElapsedMillis() > CachedPravegaNodeUri.MAX_BACKOFF_MILLIS) {
+            // Trigger a background call and refresh the cache.
+            endPointCacheMap.put(segment, new CachedPravegaNodeUri(new Timer(), getPravegaNodeUri(segment)));
+            return nodeUri.getPravegaNodeUri();
+        } else {
+            CompletableFuture<PravegaNodeUri> nodeInfo = getPravegaNodeUri(segment);
+            endPointCacheMap.put(segment, new CachedPravegaNodeUri(new Timer(), nodeInfo));
+            return nodeInfo;
+        }
+    }
+
+    @Override
+    public void updateStaleValueInCache(String segmentName, PravegaNodeUri errNodeUri) {
+        Exceptions.checkNotNullOrEmpty(segmentName, "segmentName");
+        Segment segment = Segment.fromScopedName(segmentName);
         final long requestId = requestIdGenerator.get();
-        long traceId = LoggerHelpers.traceEnter(log, "getEndpointForSegment", qualifiedSegmentName, requestId);
+        long traceId = LoggerHelpers.traceEnter(log, "updateStaleValueInCache", segmentName, errNodeUri, requestId);
+        CachedPravegaNodeUri cachedNode = getSegmentEndpointFromCache(segment);
+        if (cachedNode != null) {
+            cachedNode.getPravegaNodeUri().thenAccept(cachedNodeUri -> {
```

Comment:
If a background refresh is already in progress I am not sure we need to do anything. So it may be useful to check `cachedNode.getPravegaNodeUri().isDone()`

Reply:
This will always wait for the cachedNode to get the result and then it will check if the existing value in cache matches the error node details, only then it will trigger for the update. Please suggest if this check is to be done.

Reply:
Yes. My concern is this awaiting behavior. We previously had an issue in the read path where we found out that for some versions of Java chaining many futures off of one can lead to a O(n) time to select the next future to be run. (This was fixed in later versions) but it still can lead to a potential accidental self DOS. Imagine the scenario where the outstanding request is very slow to respond. (Do we have a timeout on it?) During that time many duplicate requests come in. This will grow memory as it sets up callbacks for each one to be run when the result comes back. (And in the case we are running the wrong JVM version, take O(n^2) time to run them all). What's more, if the request is outstanding there is a very high probability that the returned value will not be the incorrect one. After all it was just retrieved from the controller (presumably because some previous request went to the wrong host). In the worst case if does come back with some stale information, we'll just re-fetch it anyway.

Reply:
Sure, Incorporated changes.